### ISSUE TITLE
PHPStan error for `IdList::handlesIdClass`

### ISSUE THREAD

**Issue by plandolt** [digital-craftsman-de/ids#29]
I am not sure why, but phpstan reports this:
` Method App\Domain\ValueObject\IdList\CustomerIdList::handlesIdClass() should return class-string<TT of App\Domain\ValueObject\Id\CustomerId> but returns string.`

This is how my `CustomerIdList` object looks like:
```
<?php

namespace App\Domain\ValueObject\IdList;

use App\Domain\ValueObject\Id\CustomerId;
use DigitalCraftsman\Ids\ValueObject\IdList;

/** @extends IdList<CustomerId> */
final class CustomerIdList extends IdList
{
    public static function handlesIdClass(): string
    {
        return CustomerId::class;
    }
}
```

**Comment by christian-kolb** [comment]
@scuben Sorry for the late reply, I didn't get a notification. I will check why that is.

I guess this is due to the special annotation for Psalm that is necessary for static methods. I don't use PHPStan and therefore can't validate it right now, but can you replace lines 83 and 85 of the `IdList` class with `@psalm-template` and `@psalm-return` and see if it's still mentioned as an error? If it isn't then I would adapt the annotations accordingly 🙂

**Comment by plandolt** [comment]
no worries. the change to @psalm-* did not work.

**Comment by christian-kolb** [comment]
> did not work

means it still displays the same issue? It has to be a different one, because otherwise PHPStan would need to interpret the @psalm attributes which would be weird. Can you copy the new error?

Sorry, for the back and forth, otherwise I need to setup PHPStan in a living project of mine and I don't know when I would be able to do so 🙂

**Comment by christian-kolb** [comment]
Ok, you're correct. I just threw it into a project of mine (because it bugged me) and it displays exactly the same error.

It seems that the workaround that makes Psalm understand the template syntax within a static method leads to PHPStan being confused. The following syntax seems to work:

```php
/**
 * @template TT of T
 *
 * @return class-string<TT>
 * @phpstan-return class-string<T>
 */
abstract public static function handlesIdClass(): string;
```

Can you confirm that this works in your project?

**Comment by plandolt** [comment]
Sorry that I did not get to try that earlier. With this docbloc change, phpstan is working properly

**Comment by christian-kolb** [comment]
@scuben I integrated PHPStan in a project of mine on level 9 over the weekend. This way I can reproduce the problem better. I now integrated PHPStan in this package as well as I could (see linked PR). I'm validating the PR with my project and will then merge it and create a new release.
If you're able to try requiring `composer require digital-craftsman/ids:dev-fix-phpstan-issues` and tell me if this also fixes all of your issues or whether I created new ones with this 🙂
